// sanity-plugin-media/dist/components/ReduxProvider/index.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
var __spread = (this && this.__spread) || function () {
    for (var ar = [], i = 0; i < arguments.length; i++) ar = ar.concat(__read(arguments[i]));
    return ar;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var toolkit_1 = require("@reduxjs/toolkit");
var react_1 = __importStar(require("react"));
var react_redux_1 = require("react-redux");
var redux_observable_1 = require("redux-observable");
var modules_1 = require("../../modules");
var getDocumentAssetIds_1 = __importDefault(require("../../utils/getDocumentAssetIds"));
var ReduxProvider = /** @class */ (function (_super) {
    __extends(ReduxProvider, _super);
    function ReduxProvider(props) {
        var _this = _super.call(this, props) || this;
        // Initialize redux store + middleware
        var epicMiddleware = redux_observable_1.createEpicMiddleware();
        _this.store = toolkit_1.configureStore({
            reducer: modules_1.rootReducer,
            middleware: __spread([
                epicMiddleware
            ], toolkit_1.getDefaultMiddleware({
                /*
                serializableCheck: {
                  ignoredActions: [
                    assetsActions.deleteError.type,
                    uploadsActions.uploadRequest.type,
                    uploadsActions.uploadStart.type,
                  ]
                },
                */
                // TODO: remove once we're no longer storing non-serializable data in the store
                serializableCheck: false,
                thunk: false
            })),
            devTools: true,
            preloadedState: {
                selected: {
                    assets: props.selectedAssets,
                    document: props.document,
                    documentAssetIds: props.document ? getDocumentAssetIds_1.default(props.document) : []
                }
            }
        });
        epicMiddleware.run(modules_1.rootEpic);
        return _this;
    }
    ReduxProvider.prototype.render = function () {
        return react_1.default.createElement(react_redux_1.Provider, { store: this.store }, this.props.children);
    };
    return ReduxProvider;
}(react_1.Component));
exports.default = ReduxProvider;
