// sanity-plugin-media/dist/helpers/withRedux.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    Object.defineProperty(o, k2, { enumerable: true, get: function() { return m[k]; } });
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = __importStar(require("react"));
var react_redux_1 = require("react-redux");
var redux_1 = require("redux");
var developmentOnly_1 = require("redux-devtools-extension/developmentOnly");
var redux_observable_1 = require("redux-observable");
var modules_1 = require("../modules");
var assets_1 = require("../modules/assets");
function withRedux(ComposedComponent) {
    var _a;
    var displayName = ComposedComponent.displayName || ComposedComponent.name || 'Component';
    return _a = /** @class */ (function (_super) {
            __extends(WithReduxComponent, _super);
            function WithReduxComponent(props) {
                var _this = _super.call(this, props) || this;
                // Initialize redux middleware and create store.
                // Redux store isn't persisted between mounts.
                var epicMiddleware = redux_observable_1.createEpicMiddleware();
                _this.store = redux_1.createStore(modules_1.rootReducer, {
                    assets: __assign({}, assets_1.initialState),
                    // TODO: correctly type
                    // @ts-ignore
                    document: props === null || props === void 0 ? void 0 : props.document,
                    // TODO: correctly type
                    // @ts-ignore
                    selectedAssets: props === null || props === void 0 ? void 0 : props.selectedAssets
                }, developmentOnly_1.composeWithDevTools(redux_1.applyMiddleware(epicMiddleware)));
                epicMiddleware.run(modules_1.rootEpic);
                return _this;
            }
            WithReduxComponent.prototype.render = function () {
                return (react_1.default.createElement(react_redux_1.Provider, { store: this.store },
                    react_1.default.createElement(ComposedComponent, __assign({}, this.props))));
            };
            return WithReduxComponent;
        }(react_1.Component)),
        _a.displayName = "withRedux(" + displayName + ")",
        _a;
}
exports.default = withRedux;
