// sanity-plugin-media/dist/components/DropzoneTest/index.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = __importDefault(require("react"));
var react_dropzone_1 = require("react-dropzone");
var DropzoneTest = function (props) {
    var _a = react_dropzone_1.useDropzone({
        // getFilesFromEvent: handleFileGetter,
        noClick: true
        // onDrop: handleDrop
    }), getRootProps = _a.getRootProps, getInputProps = _a.getInputProps, isDragActive = _a.isDragActive, open = _a.open;
    return (react_1.default.createElement("div", __assign({ style: {
            background: 'blue',
            color: 'white',
            width: '100%',
            height: '100%',
            position: 'absolute',
            zIndex: 2
        } }, getRootProps()),
        react_1.default.createElement("input", __assign({}, getInputProps())),
        react_1.default.createElement("div", null, "dropzone test"),
        react_1.default.createElement("div", null, isDragActive.toString())));
};
exports.default = DropzoneTest;
