// sanity-plugin-vercel-deploy/dist/deploy-log.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _slicedToArray2 = _interopRequireDefault(require("@babel/runtime/helpers/slicedToArray"));

var _ui = require("@sanity/ui");

var _axios = _interopRequireDefault(require("axios"));

var _react = _interopRequireWildcard(require("react"));

var _timeAgo = _interopRequireDefault(require("./time-ago"));

var _deployLog = _interopRequireDefault(require("./deploy-log.css"));

var _status = _interopRequireWildcard(require("./status"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { "default": obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj["default"] = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

var DeployLog = function DeployLog(_ref) {
  var _state$deployments;

  var vercelProject = _ref.vercelProject,
      vercelToken = _ref.vercelToken,
      vercelTeam = _ref.vercelTeam;

  var _useState = (0, _react.useState)({}),
      _useState2 = (0, _slicedToArray2["default"])(_useState, 2),
      state = _useState2[0],
      setState = _useState2[1];

  (0, _react.useEffect)(function () {
    if (!vercelProject) {
      return;
    }

    setState({
      loading: true
    });
    var options = {
      method: 'GET',
      headers: {
        'content-type': 'application/json',
        Authorization: "Bearer ".concat(vercelToken)
      },
      url: "https://api.vercel.com/v5/now/deployments?projectId=".concat(vercelProject, "&limit=5").concat(vercelTeam !== null && vercelTeam !== void 0 && vercelTeam.id ? "&teamId=".concat(vercelTeam === null || vercelTeam === void 0 ? void 0 : vercelTeam.id) : '')
    };
    (0, _axios["default"])(options).then(function (_ref2) {
      var data = _ref2.data;
      setState({
        deployments: data.deployments,
        loading: false,
        error: false
      });
    })["catch"](function (e) {
      setState({
        error: true,
        loading: false
      });
      console.warn(e);
    });
  }, [vercelProject]);

  if (state.loading) {
    return /*#__PURE__*/_react["default"].createElement(_ui.Flex, {
      align: "center",
      justify: "center"
    }, /*#__PURE__*/_react["default"].createElement(_ui.Spinner, {
      muted: true
    }));
  }

  if (state.error) {
    return /*#__PURE__*/_react["default"].createElement(_ui.Card, {
      padding: [3, 3, 4],
      radius: 2,
      shadow: 1,
      tone: "critical"
    }, /*#__PURE__*/_react["default"].createElement(_ui.Text, {
      size: [2, 2, 3]
    }, "Could not load deployments for ", vercelProject));
  }

  return /*#__PURE__*/_react["default"].createElement(_ui.Box, {
    as: "table",
    className: _deployLog["default"].table
  }, /*#__PURE__*/_react["default"].createElement(_ui.Box, {
    as: "thead",
    style: {
      display: 'table-header-group'
    }
  }, /*#__PURE__*/_react["default"].createElement("tr", null, /*#__PURE__*/_react["default"].createElement("th", null, "Deployment"), /*#__PURE__*/_react["default"].createElement("th", null, "State"), /*#__PURE__*/_react["default"].createElement("th", null, "Commit"), /*#__PURE__*/_react["default"].createElement("th", null, "Time"), /*#__PURE__*/_react["default"].createElement("th", null, "Creator"))), /*#__PURE__*/_react["default"].createElement(_ui.Box, {
    as: "tbody",
    style: {
      display: 'table-row-group'
    }
  }, (_state$deployments = state.deployments) === null || _state$deployments === void 0 ? void 0 : _state$deployments.map(function (deployment) {
    var _deployment$meta, _deployment$meta2, _deployment$creator, _deployment$creator2, _deployment$creator3;

    return /*#__PURE__*/_react["default"].createElement("tr", {
      as: "tr",
      key: deployment.uid
    }, /*#__PURE__*/_react["default"].createElement("td", null, /*#__PURE__*/_react["default"].createElement("a", {
      href: "https://".concat(deployment.url),
      target: "_blank"
    }, deployment.url)), /*#__PURE__*/_react["default"].createElement("td", null, /*#__PURE__*/_react["default"].createElement(_status["default"], {
      status: deployment.state
    }, (0, _status.titleCase)(deployment.state))), /*#__PURE__*/_react["default"].createElement("td", null, /*#__PURE__*/_react["default"].createElement("div", null, (_deployment$meta = deployment.meta) === null || _deployment$meta === void 0 ? void 0 : _deployment$meta.githubCommitRef), /*#__PURE__*/_react["default"].createElement("small", {
      className: _deployLog["default"].commit
    }, (_deployment$meta2 = deployment.meta) === null || _deployment$meta2 === void 0 ? void 0 : _deployment$meta2.githubCommitMessage)), /*#__PURE__*/_react["default"].createElement("td", null, /*#__PURE__*/_react["default"].createElement(_timeAgo["default"], {
      date: deployment.created
    })), /*#__PURE__*/_react["default"].createElement("td", null, /*#__PURE__*/_react["default"].createElement(_ui.Tooltip, {
      content: /*#__PURE__*/_react["default"].createElement(_ui.Box, {
        padding: 2
      }, /*#__PURE__*/_react["default"].createElement(_ui.Text, {
        muted: true,
        size: 1
      }, (_deployment$creator = deployment.creator) === null || _deployment$creator === void 0 ? void 0 : _deployment$creator.username)),
      fallbackPlacements: ['right', 'left'],
      placement: "top"
    }, /*#__PURE__*/_react["default"].createElement(_ui.Avatar, {
      alt: (_deployment$creator2 = deployment.creator) === null || _deployment$creator2 === void 0 ? void 0 : _deployment$creator2.username,
      src: "https://vercel.com/api/www/avatar/".concat((_deployment$creator3 = deployment.creator) === null || _deployment$creator3 === void 0 ? void 0 : _deployment$creator3.uid, "?&s=48"),
      size: 1,
      style: {
        margin: 'auto'
      }
    }))));
  })));
};

var _default = DeployLog;
exports["default"] = _default;
