// sanity-plugin-media/dist/components/Dropzone/index.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var react_1 = __importDefault(require("react"));
var react_dropzone_1 = require("react-dropzone");
var Dropzone = function (props) {
    var children = props.children;
    var handleDrop = function (acceptedFiles) {
        console.log('ON DROP');
        console.log('acceptedFiles', acceptedFiles);
    };
    var _a = react_dropzone_1.useDropzone({ onDrop: handleDrop }), getRootProps = _a.getRootProps, getInputProps = _a.getInputProps, isDragActive = _a.isDragActive;
    /*
    return (
      <div
        style={{
          background: 'red',
          color: 'black',
          height: '100%',
          padding: '10px',
          // pointerEvents: isDragActive ? 'auto' : 'none',
          position: 'absolute',
          right: 0,
          top: 0,
          width: '100%',
          zIndex: 1
          // zIndex: isDragActive ? 1 : 0
        }}
      >
        yoo
      </div>
    )
    */
    return (react_1.default.createElement("div", __assign({}, getRootProps(), { style: {
            background: 'red',
            color: 'white',
            // height: '100%',
            height: '500px',
            opacity: isDragActive ? 1 : 0.25,
            padding: '10px',
            // pointerEvents: isDragActive ? 'auto' : 'none',
            position: 'absolute',
            right: 0,
            top: 0,
            // width: '100%',
            width: '500px',
            zIndex: 1
            // zIndex: isDragActive ? 1 : 0
        } }),
        "Drop zone",
        children));
};
exports.default = Dropzone;
