// sanity-plugin-media/dist/modules/notifications/notificationsSlice.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.notificationsTagUpdateCompleteEpic = exports.notificationsTagDeleteCompleteEpic = exports.notificationsTagCreateCompleteEpic = exports.notificationsGenericErrorEpic = exports.notificationsAssetsUpdateCompleteEpic = exports.notificationsAssetsTagsRemoveCompleteEpic = exports.notificationsAssetsTagsAddCompleteEpic = exports.notificationsAssetsDeleteErrorEpic = exports.notificationsAssetsDeleteCompleteEpic = void 0;
var toolkit_1 = require("@reduxjs/toolkit");
var pluralize_1 = __importDefault(require("pluralize"));
var redux_observable_1 = require("redux-observable");
var rxjs_1 = require("rxjs");
var operators_1 = require("rxjs/operators");
var assets_1 = require("../assets");
var tagsSlice_1 = require("../tags/tagsSlice");
var initialState = {
    items: []
};
var notificationsSlice = toolkit_1.createSlice({
    name: 'notifications',
    initialState: initialState,
    reducers: {
        add: function (state, action) {
            var _a = action.payload, asset = _a.asset, status = _a.status, subtitle = _a.subtitle, title = _a.title;
            state.items.push({
                asset: asset,
                status: status,
                subtitle: subtitle,
                title: title
            });
        }
    }
});
var notificationsAssetsDeleteCompleteEpic = function (action$) {
    return action$.pipe(operators_1.filter(assets_1.deleteComplete.match), operators_1.bufferTime(1000), operators_1.filter(function (actions) { return actions.length > 0; }), operators_1.mergeMap(function (actions) {
        var deletedCount = actions.length;
        return rxjs_1.of(notificationsSlice.actions.add({
            status: 'success',
            title: deletedCount + " " + pluralize_1.default('asset', deletedCount) + " deleted"
        }));
    }));
};
exports.notificationsAssetsDeleteCompleteEpic = notificationsAssetsDeleteCompleteEpic;
var notificationsAssetsDeleteErrorEpic = function (action$) {
    return action$.pipe(operators_1.filter(assets_1.deleteError.match), operators_1.bufferTime(1000), operators_1.filter(function (actions) { return actions.length > 0; }), operators_1.mergeMap(function (actions) {
        var count = actions.length;
        return rxjs_1.of(
        // TODO: add error message if count === 1
        notificationsSlice.actions.add({
            status: 'error',
            title: "Unable to delete " + count + " " + pluralize_1.default('asset', count)
        }));
    }));
};
exports.notificationsAssetsDeleteErrorEpic = notificationsAssetsDeleteErrorEpic;
var notificationsAssetsTagsAddCompleteEpic = function (action$) {
    return action$.pipe(operators_1.filter(assets_1.tagsAddComplete.match), operators_1.mergeMap(function (action) {
        var _a, _b;
        var count = (_b = (_a = action === null || action === void 0 ? void 0 : action.payload) === null || _a === void 0 ? void 0 : _a.assets) === null || _b === void 0 ? void 0 : _b.length;
        return rxjs_1.of(notificationsSlice.actions.add({
            status: 'success',
            title: "Tag added to " + count + " " + pluralize_1.default('asset', count)
        }));
    }));
};
exports.notificationsAssetsTagsAddCompleteEpic = notificationsAssetsTagsAddCompleteEpic;
var notificationsAssetsTagsRemoveCompleteEpic = function (action$) {
    return action$.pipe(operators_1.filter(assets_1.tagsRemoveComplete.match), operators_1.mergeMap(function (action) {
        var _a, _b;
        var count = (_b = (_a = action === null || action === void 0 ? void 0 : action.payload) === null || _a === void 0 ? void 0 : _a.assets) === null || _b === void 0 ? void 0 : _b.length;
        return rxjs_1.of(notificationsSlice.actions.add({
            status: 'success',
            title: "Tag removed from " + count + " " + pluralize_1.default('asset', count)
        }));
    }));
};
exports.notificationsAssetsTagsRemoveCompleteEpic = notificationsAssetsTagsRemoveCompleteEpic;
var notificationsAssetsUpdateCompleteEpic = function (action$) {
    return action$.pipe(operators_1.filter(assets_1.updateComplete.match), operators_1.mergeMap(function () {
        return rxjs_1.of(notificationsSlice.actions.add({
            status: 'success',
            title: "Asset updated"
        }));
    }));
};
exports.notificationsAssetsUpdateCompleteEpic = notificationsAssetsUpdateCompleteEpic;
var notificationsGenericErrorEpic = function (action$) {
    return action$.pipe(redux_observable_1.ofType(assets_1.fetchError.type, assets_1.updateError.type, tagsSlice_1.createError.type, tagsSlice_1.fetchError.type, tagsSlice_1.updateError.type), operators_1.mergeMap(function (action) {
        var _a;
        var error = (_a = action.payload) === null || _a === void 0 ? void 0 : _a.error;
        return rxjs_1.of(notificationsSlice.actions.add({
            status: 'error',
            title: "An error occured: " + error.message
        }));
    }));
};
exports.notificationsGenericErrorEpic = notificationsGenericErrorEpic;
var notificationsTagCreateCompleteEpic = function (action$) {
    return action$.pipe(operators_1.filter(tagsSlice_1.createComplete.match), operators_1.mergeMap(function () {
        return rxjs_1.of(notificationsSlice.actions.add({
            status: 'success',
            title: "Tag created"
        }));
    }));
};
exports.notificationsTagCreateCompleteEpic = notificationsTagCreateCompleteEpic;
var notificationsTagDeleteCompleteEpic = function (action$) {
    return action$.pipe(operators_1.filter(tagsSlice_1.deleteComplete.match), operators_1.mergeMap(function () {
        return rxjs_1.of(notificationsSlice.actions.add({
            status: 'success',
            title: "Tag deleted"
        }));
    }));
};
exports.notificationsTagDeleteCompleteEpic = notificationsTagDeleteCompleteEpic;
var notificationsTagUpdateCompleteEpic = function (action$) {
    return action$.pipe(operators_1.filter(tagsSlice_1.updateComplete.match), operators_1.mergeMap(function () {
        return rxjs_1.of(notificationsSlice.actions.add({
            status: 'success',
            title: "Tag updated"
        }));
    }));
};
exports.notificationsTagUpdateCompleteEpic = notificationsTagUpdateCompleteEpic;
exports.default = notificationsSlice.reducer;
