// sanity-plugin-media/dist/components/PanelHeader/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var color_1 = require("@sanity/color");
var icons_1 = require("@sanity/icons");
var ui_1 = require("@sanity/ui");
var react_1 = __importDefault(require("react"));
var react_redux_1 = require("react-redux");
var constants_1 = require("../../constants");
var useTypedSelector_1 = __importDefault(require("../../hooks/useTypedSelector"));
var dialog_1 = require("../../modules/dialog");
var Tag_1 = __importDefault(require("../Tag"));
var PanelHeader = function (props) {
    var _a, _b;
    var actions = props.actions, allowCreate = props.allowCreate, light = props.light, tags = props.tags, title = props.title;
    // Redux
    var dispatch = react_redux_1.useDispatch();
    var tagsCreating = useTypedSelector_1.default(function (state) { return state.tags.creating; });
    var tagsFetching = useTypedSelector_1.default(function (state) { return state.tags.fetching; });
    // Callbacks
    var handleTagCreate = function () {
        dispatch(dialog_1.dialogActions.showTagCreate());
    };
    return (react_1.default.createElement(react_1.default.Fragment, null,
        react_1.default.createElement(ui_1.Flex, { align: "center", justify: "space-between", paddingLeft: 3, style: {
                background: light ? (_a = color_1.hues.gray) === null || _a === void 0 ? void 0 : _a[900].hex : color_1.black.hex,
                borderBottom: "1px solid " + ((_b = color_1.hues.gray) === null || _b === void 0 ? void 0 : _b[900].hex),
                height: constants_1.PANEL_HEIGHT + "px",
                position: 'sticky',
                top: 0,
                zIndex: 1 // force stacking context
            } },
            react_1.default.createElement(ui_1.Inline, { space: 2 },
                react_1.default.createElement(ui_1.Label, { size: 0 }, title),
                tagsFetching && (react_1.default.createElement(ui_1.Label, { size: 0, style: { opacity: 0.3 } }, "Loading..."))),
            allowCreate && (react_1.default.createElement(ui_1.Box, { marginRight: 1 },
                react_1.default.createElement(ui_1.Button, { disabled: tagsCreating, fontSize: 1, icon: icons_1.ComposeIcon, mode: "bleed", onClick: handleTagCreate, style: {
                        background: 'transparent',
                        boxShadow: 'none'
                    } })))),
        tags && (react_1.default.createElement(ui_1.Box, { paddingLeft: 3, paddingRight: 2 }, tags === null || tags === void 0 ? void 0 : tags.map(function (tagItem) {
            var _a;
            return (react_1.default.createElement(Tag_1.default, { actions: actions, key: (_a = tagItem === null || tagItem === void 0 ? void 0 : tagItem.tag) === null || _a === void 0 ? void 0 : _a._id, tag: tagItem }));
        })))));
};
exports.default = PanelHeader;
