// sanity-plugin-media/dist/components/CardUpload copy/index.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
var ui_1 = require("@sanity/ui");
var filesize_1 = __importDefault(require("filesize"));
var react_1 = __importDefault(require("react"));
var constants_1 = require("../../constants");
var Image_1 = __importDefault(require("../Image"));
var CardUpload = function (props) {
    var item = props.item, style = props.style;
    var fileSize = filesize_1.default(item.size, { base: 10, round: 0 });
    var status;
    /*
    if (item.status === 'complete') {
      status = 'Complete'
    }
    */
    if (['complete', 'uploading'].includes(item.status)) {
        status = Math.round((item === null || item === void 0 ? void 0 : item.percent) || 0) + "%";
    }
    if (item.status === 'queued') {
        status = 'Queued';
    }
    return (react_1.default.createElement(ui_1.Flex, { direction: "column", style: __assign(__assign({}, style), { border: '1px solid transparent' }) },
        react_1.default.createElement(ui_1.Box, { flex: 1, style: { position: 'relative' } },
            (item === null || item === void 0 ? void 0 : item.objectUrl) && (react_1.default.createElement(Image_1.default, { draggable: false, src: item.objectUrl, style: {
                    opacity: 0.15
                } })),
            react_1.default.createElement(ui_1.Flex, { align: "center", justify: "center", style: {
                    height: '100%',
                    left: 0,
                    position: 'absolute',
                    top: 0,
                    width: '100%'
                } },
                react_1.default.createElement(ui_1.Text, { size: 1, style: { opacity: item.status === 'queued' ? 0.25 : 1 } }, status))),
        react_1.default.createElement(ui_1.Flex, { align: "center", paddingX: 1, style: { height: constants_1.PANEL_HEIGHT + "px" } },
            react_1.default.createElement(ui_1.Box, null,
                react_1.default.createElement(ui_1.Text, { size: 0, textOverflow: "ellipsis" },
                    "Uploading ",
                    item.name,
                    " (",
                    fileSize,
                    ")",
                    ' ')))));
};
exports.default = CardUpload;
